File: Alcoholtracker/Services/MixerDatabase.swift
import Foundation

// MARK: - MixerDatabase

enum MixerDatabase {

    static let all: [Mixer] = sodas + bitters + energyDrinks + juices + waters + creams + teas + other

    static func entries(for category: MixerCategory) -> [Mixer] {
        all.filter { $0.category == category }
    }

    static func search(_ query: String) -> [Mixer] {
        guard !query.trimmingCharacters(in: .whitespaces).isEmpty else { return all }
        let q = query.lowercased()
        return all.filter { $0.name.lowercased().contains(q) }
    }

    static func grouped() -> [(MixerCategory, [Mixer])] {
        let grouped = Dictionary(grouping: all, by: { $0.category })
        return MixerCategory.allCases.compactMap { cat in
            guard let mixers = grouped[cat] else { return nil }
            return (cat, mixers.sorted { $0.name < $1.name })
        }
    }

    // MARK: Softdrinks (10 Eintraege)

    private static let sodas: [Mixer] = [
        Mixer(name: "Coca-Cola",        category: .soda, caloriesPer100ml: 41, waterContentPercent: 89,  icon: "cup.and.saucer.fill"),
        Mixer(name: "Coca-Cola Zero",   category: .soda, caloriesPer100ml: 0,  waterContentPercent: 99,  icon: "cup.and.saucer.fill"),
        Mixer(name: "Pepsi",            category: .soda, caloriesPer100ml: 42, waterContentPercent: 89,  icon: "cup.and.saucer.fill"),
        Mixer(name: "Pepsi Max",        category: .soda, caloriesPer100ml: 0,  waterContentPercent: 99,  icon: "cup.and.saucer.fill"),
        Mixer(name: "Sprite",           category: .soda, caloriesPer100ml: 38, waterContentPercent: 90,  icon: "cup.and.saucer.fill"),
        Mixer(name: "Fanta Orange",     category: .soda, caloriesPer100ml: 40, waterContentPercent: 89,  icon: "cup.and.saucer.fill"),
        Mixer(name: "7Up",              category: .soda, caloriesPer100ml: 38, waterContentPercent: 90,  icon: "cup.and.saucer.fill"),
        Mixer(name: "Spezi",            category: .soda, caloriesPer100ml: 40, waterContentPercent: 88,  icon: "cup.and.saucer.fill"),
        Mixer(name: "Limonade Zitrone", category: .soda, caloriesPer100ml: 38, waterContentPercent: 90,  icon: "cup.and.saucer.fill"),
        Mixer(name: "Club Soda",        category: .soda, caloriesPer100ml: 0,  waterContentPercent: 100, icon: "cup.and.saucer.fill"),
    ]

    // MARK: Bitter (5 Eintraege)

    private static let bitters: [Mixer] = [
        Mixer(name: "Tonic Water",      category: .bitter, caloriesPer100ml: 35, waterContentPercent: 91, icon: "drop.fill"),
        Mixer(name: "Bitter Lemon",     category: .bitter, caloriesPer100ml: 37, waterContentPercent: 91, icon: "drop.fill"),
        Mixer(name: "Ginger Ale",       category: .bitter, caloriesPer100ml: 35, waterContentPercent: 90, icon: "drop.fill"),
        Mixer(name: "Ginger Beer",      category: .bitter, caloriesPer100ml: 42, waterContentPercent: 89, icon: "drop.fill"),
        Mixer(name: "Schweppes Russian",category: .bitter, caloriesPer100ml: 36, waterContentPercent: 91, icon: "drop.fill"),
    ]

    // MARK: Energy Drinks (7 Eintraege)

    private static let energyDrinks: [Mixer] = [
        Mixer(name: "Red Bull",            category: .energy, caloriesPer100ml: 46, waterContentPercent: 88, icon: "bolt.fill"),
        Mixer(name: "Red Bull Sugarfree",  category: .energy, caloriesPer100ml: 4,  waterContentPercent: 99, icon: "bolt.fill"),
        Mixer(name: "Monster Energy",      category: .energy, caloriesPer100ml: 46, waterContentPercent: 88, icon: "bolt.fill"),
        Mixer(name: "Monster Zero",        category: .energy, caloriesPer100ml: 4,  waterContentPercent: 99, icon: "bolt.fill"),
        Mixer(name: "Rockstar",            category: .energy, caloriesPer100ml: 50, waterContentPercent: 87, icon: "bolt.fill"),
        Mixer(name: "Effect Energy",       category: .energy, caloriesPer100ml: 47, waterContentPercent: 87, icon: "bolt.fill"),
        Mixer(name: "Burn Energy Drink",   category: .energy, caloriesPer100ml: 44, waterContentPercent: 88, icon: "bolt.fill"),
    ]

    // MARK: Saefte (13 Eintraege)

    private static let juices: [Mixer] = [
        Mixer(name: "Orangensaft",    category: .juice, caloriesPer100ml: 45, waterContentPercent: 88, icon: "leaf.fill"),
        Mixer(name: "Apfelsaft",      category: .juice, caloriesPer100ml: 47, waterContentPercent: 87, icon: "leaf.fill"),
        Mixer(name: "Cranberrysaft",  category: .juice, caloriesPer100ml: 46, waterContentPercent: 87, icon: "leaf.fill"),
        Mixer(name: "Ananassaft",     category: .juice, caloriesPer100ml: 53, waterContentPercent: 86, icon: "leaf.fill"),
        Mixer(name: "Grapefruitsaft", category: .juice, caloriesPer100ml: 38, waterContentPercent: 90, icon: "leaf.fill"),
        Mixer(name: "Limettensaft",   category: .juice, caloriesPer100ml: 25, waterContentPercent: 93, icon: "leaf.fill"),
        Mixer(name: "Zitronensaft",   category: .juice, caloriesPer100ml: 22, waterContentPercent: 93, icon: "leaf.fill"),
        Mixer(name: "Tomatensaft",    category: .juice, caloriesPer100ml: 17, waterContentPercent: 94, icon: "leaf.fill"),
        Mixer(name: "Maracujasaft",   category: .juice, caloriesPer100ml: 51, waterContentPercent: 86, icon: "leaf.fill"),
        Mixer(name: "Traubensaft",    category: .juice, caloriesPer100ml: 60, waterContentPercent: 84, icon: "leaf.fill"),
        Mixer(name: "Kokoswasser",    category: .juice, caloriesPer100ml: 19, waterContentPercent: 95, icon: "leaf.fill"),
        Mixer(name: "Pfirsichsaft",   category: .juice, caloriesPer100ml: 49, waterContentPercent: 88, icon: "leaf.fill"),
        Mixer(name: "Multivitamin",   category: .juice, caloriesPer100ml: 50, waterContentPercent: 87, icon: "leaf.fill"),
    ]

    // MARK: Wasser (3 Eintraege)

    private static let waters: [Mixer] = [
        Mixer(name: "Mineralwasser", category: .water, caloriesPer100ml: 0, waterContentPercent: 100, icon: "drop"),
        Mixer(name: "Soda Water",    category: .water, caloriesPer100ml: 0, waterContentPercent: 100, icon: "drop"),
        Mixer(name: "Sprudelwasser", category: .water, caloriesPer100ml: 0, waterContentPercent: 100, icon: "drop"),
    ]

    // MARK: Cremig (5 Eintraege)

    private static let creams: [Mixer] = [
        Mixer(name: "Sahne",       category: .cream, caloriesPer100ml: 340, waterContentPercent: 57, icon: "drop.circle.fill"),
        Mixer(name: "Kokosmilch",  category: .cream, caloriesPer100ml: 180, waterContentPercent: 66, icon: "drop.circle.fill"),
        Mixer(name: "Vollmilch",   category: .cream, caloriesPer100ml: 61,  waterContentPercent: 88, icon: "drop.circle.fill"),
        Mixer(name: "Mandelmilch", category: .cream, caloriesPer100ml: 24,  waterContentPercent: 96, icon: "drop.circle.fill"),
        Mixer(name: "Haferdrink",  category: .cream, caloriesPer100ml: 46,  waterContentPercent: 89, icon: "drop.circle.fill"),
    ]

    // MARK: Tee (4 Eintraege)

    private static let teas: [Mixer] = [
        Mixer(name: "Eistee Pfirsich", category: .tea, caloriesPer100ml: 30, waterContentPercent: 92, icon: "cup.and.saucer"),
        Mixer(name: "Eistee Zitrone",  category: .tea, caloriesPer100ml: 28, waterContentPercent: 92, icon: "cup.and.saucer"),
        Mixer(name: "Schwarzer Tee",   category: .tea, caloriesPer100ml: 1,  waterContentPercent: 99, icon: "cup.and.saucer"),
        Mixer(name: "Gruener Tee",     category: .tea, caloriesPer100ml: 1,  waterContentPercent: 99, icon: "cup.and.saucer"),
    ]

    // MARK: Sirup und Sonstiges (8 Eintraege)

    private static let other: [Mixer] = [
        Mixer(name: "Grenadine",       category: .other, caloriesPer100ml: 280, waterContentPercent: 40, icon: "star.fill"),
        Mixer(name: "Zuckersirup",     category: .other, caloriesPer100ml: 290, waterContentPercent: 30, icon: "star.fill"),
        Mixer(name: "Agavensirup",     category: .other, caloriesPer100ml: 310, waterContentPercent: 20, icon: "star.fill"),
        Mixer(name: "Limettenkordial", category: .other, caloriesPer100ml: 250, waterContentPercent: 40, icon: "star.fill"),
        Mixer(name: "Maracujasirup",   category: .other, caloriesPer100ml: 270, waterContentPercent: 35, icon: "star.fill"),
        Mixer(name: "Ingwersirup",     category: .other, caloriesPer100ml: 295, waterContentPercent: 28, icon: "star.fill"),
        Mixer(name: "Erdbeersirup",    category: .other, caloriesPer100ml: 265, waterContentPercent: 37, icon: "star.fill"),
        Mixer(name: "Espresso",        category: .other, caloriesPer100ml: 5,   waterContentPercent: 98, icon: "cup.and.saucer.fill"),
    ]
}

File: android/bac/src/main/kotlin/de/tipau/promille/bac/MixerDatabase.kt
package de.tipau.promille.bac

/**
 * Non-alcoholic component of a mixed drink. It carries no alcohol, so it only
 * affects calories and the hydration model, never the permille number.
 */
enum class MixerCategory(
    val raw: String,
    val germanName: String,
    val symbolName: String
) {
    SODA("soda", "Softdrink", "cup.and.saucer.fill"),
    BITTER("bitter", "Bitter", "drop.fill"),
    ENERGY("energy", "Energy", "bolt.fill"),
    JUICE("juice", "Saft", "leaf.fill"),
    WATER("water", "Wasser", "drop"),
    CREAM("cream", "Cremig", "drop.circle.fill"),
    TEA("tea", "Tee", "cup.and.saucer"),
    OTHER("other", "Sonstiges", "star.fill");

    companion object {
        fun from(raw: String): MixerCategory = entries.firstOrNull { it.raw == raw } ?: OTHER
    }
}

data class Mixer(
    val name: String,
    val category: MixerCategory,
    val caloriesPer100ml: Int,
    val waterContentPercent: Double,
    val icon: String
) {
    val id: String get() = name
}

/**
 * Built-in mixer catalog, generated from Services/MixerDatabase.swift. The values
 * feed calories and hydration, so keep both sides in step.
 */
object MixerDatabase {

    fun entries(category: MixerCategory): List<Mixer> = ALL.filter { it.category == category }

    /** A blank query returns everything, matching the iOS picker. */
    fun search(query: String): List<Mixer> {
        if (query.trim().isEmpty()) return ALL
        val q = query.lowercase()
        return ALL.filter { it.name.lowercase().contains(q) }
    }

    /** Categories in declaration order, names sorted inside each one. */
    fun grouped(): List<Pair<MixerCategory, List<Mixer>>> {
        val byCategory = ALL.groupBy { it.category }
        return MixerCategory.entries.mapNotNull { cat ->
            byCategory[cat]?.let { cat to it.sortedBy { m -> m.name } }
        }
    }

    val ALL: List<Mixer> = listOf(
        Mixer("Coca-Cola", MixerCategory.SODA, 41, 89.0, "cup.and.saucer.fill"),
        Mixer("Coca-Cola Zero", MixerCategory.SODA, 0, 99.0, "cup.and.saucer.fill"),
        Mixer("Pepsi", MixerCategory.SODA, 42, 89.0, "cup.and.saucer.fill"),
        Mixer("Pepsi Max", MixerCategory.SODA, 0, 99.0, "cup.and.saucer.fill"),
        Mixer("Sprite", MixerCategory.SODA, 38, 90.0, "cup.and.saucer.fill"),
        Mixer("Fanta Orange", MixerCategory.SODA, 40, 89.0, "cup.and.saucer.fill"),
        Mixer("7Up", MixerCategory.SODA, 38, 90.0, "cup.and.saucer.fill"),
        Mixer("Spezi", MixerCategory.SODA, 40, 88.0, "cup.and.saucer.fill"),
        Mixer("Limonade Zitrone", MixerCategory.SODA, 38, 90.0, "cup.and.saucer.fill"),
        Mixer("Club Soda", MixerCategory.SODA, 0, 100.0, "cup.and.saucer.fill"),
        Mixer("Tonic Water", MixerCategory.BITTER, 35, 91.0, "drop.fill"),
        Mixer("Bitter Lemon", MixerCategory.BITTER, 37, 91.0, "drop.fill"),
        Mixer("Ginger Ale", MixerCategory.BITTER, 35, 90.0, "drop.fill"),
        Mixer("Ginger Beer", MixerCategory.BITTER, 42, 89.0, "drop.fill"),
        Mixer("Schweppes Russian", MixerCategory.BITTER, 36, 91.0, "drop.fill"),
        Mixer("Red Bull", MixerCategory.ENERGY, 46, 88.0, "bolt.fill"),
        Mixer("Red Bull Sugarfree", MixerCategory.ENERGY, 4, 99.0, "bolt.fill"),
        Mixer("Monster Energy", MixerCategory.ENERGY, 46, 88.0, "bolt.fill"),
        Mixer("Monster Zero", MixerCategory.ENERGY, 4, 99.0, "bolt.fill"),
        Mixer("Rockstar", MixerCategory.ENERGY, 50, 87.0, "bolt.fill"),
        Mixer("Effect Energy", MixerCategory.ENERGY, 47, 87.0, "bolt.fill"),
        Mixer("Burn Energy Drink", MixerCategory.ENERGY, 44, 88.0, "bolt.fill"),
        Mixer("Orangensaft", MixerCategory.JUICE, 45, 88.0, "leaf.fill"),
        Mixer("Apfelsaft", MixerCategory.JUICE, 47, 87.0, "leaf.fill"),
        Mixer("Cranberrysaft", MixerCategory.JUICE, 46, 87.0, "leaf.fill"),
        Mixer("Ananassaft", MixerCategory.JUICE, 53, 86.0, "leaf.fill"),
        Mixer("Grapefruitsaft", MixerCategory.JUICE, 38, 90.0, "leaf.fill"),
        Mixer("Limettensaft", MixerCategory.JUICE, 25, 93.0, "leaf.fill"),
        Mixer("Zitronensaft", MixerCategory.JUICE, 22, 93.0, "leaf.fill"),
        Mixer("Tomatensaft", MixerCategory.JUICE, 17, 94.0, "leaf.fill"),
        Mixer("Maracujasaft", MixerCategory.JUICE, 51, 86.0, "leaf.fill"),
        Mixer("Traubensaft", MixerCategory.JUICE, 60, 84.0, "leaf.fill"),
        Mixer("Kokoswasser", MixerCategory.JUICE, 19, 95.0, "leaf.fill"),
        Mixer("Pfirsichsaft", MixerCategory.JUICE, 49, 88.0, "leaf.fill"),
        Mixer("Multivitamin", MixerCategory.JUICE, 50, 87.0, "leaf.fill"),
        Mixer("Mineralwasser", MixerCategory.WATER, 0, 100.0, "drop"),
        Mixer("Soda Water", MixerCategory.WATER, 0, 100.0, "drop"),
        Mixer("Sprudelwasser", MixerCategory.WATER, 0, 100.0, "drop"),
        Mixer("Sahne", MixerCategory.CREAM, 340, 57.0, "drop.circle.fill"),
        Mixer("Kokosmilch", MixerCategory.CREAM, 180, 66.0, "drop.circle.fill"),
        Mixer("Vollmilch", MixerCategory.CREAM, 61, 88.0, "drop.circle.fill"),
        Mixer("Mandelmilch", MixerCategory.CREAM, 24, 96.0, "drop.circle.fill"),
        Mixer("Haferdrink", MixerCategory.CREAM, 46, 89.0, "drop.circle.fill"),
        Mixer("Eistee Pfirsich", MixerCategory.TEA, 30, 92.0, "cup.and.saucer"),
        Mixer("Eistee Zitrone", MixerCategory.TEA, 28, 92.0, "cup.and.saucer"),
        Mixer("Schwarzer Tee", MixerCategory.TEA, 1, 99.0, "cup.and.saucer"),
        Mixer("Gruener Tee", MixerCategory.TEA, 1, 99.0, "cup.and.saucer"),
        Mixer("Grenadine", MixerCategory.OTHER, 280, 40.0, "star.fill"),
        Mixer("Zuckersirup", MixerCategory.OTHER, 290, 30.0, "star.fill"),
        Mixer("Agavensirup", MixerCategory.OTHER, 310, 20.0, "star.fill"),
        Mixer("Limettenkordial", MixerCategory.OTHER, 250, 40.0, "star.fill"),
        Mixer("Maracujasirup", MixerCategory.OTHER, 270, 35.0, "star.fill"),
        Mixer("Ingwersirup", MixerCategory.OTHER, 295, 28.0, "star.fill"),
        Mixer("Erdbeersirup", MixerCategory.OTHER, 265, 37.0, "star.fill"),
        Mixer("Espresso", MixerCategory.OTHER, 5, 98.0, "cup.and.saucer.fill"),
    )
}
